### ISSUE TITLE
gp(inference): implement VariationalInference, ConjugateVI, and Wave 3 likelihood families

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#30]
Source draft ID: `PYX-25`

The wave-level design context is copied below so this issue stays self-contained for another implementing agent.

## Shared Context
Wave 3 is the first mixed GP-and-NN wave. The relevant source docs were `features/gp/variational_families.md`, `features/gp/inference_strategies.md`, `features/gp/likelihoods.md`, `features/nn/edward2_layers.md`, `features/nn/random_features.md`, `api/nn/layers.md`, and the NN/GP example notes.

This wave intentionally opens two tracks at once:

- GP track: sparse variational inference, guide families, and first non-conjugate likelihoods.
- NN track: foundational uncertainty-aware layers plus the base random-feature surface.

The shared dependency is still `pyrox._core`. Both tracks expect `PyroModule`, `Parameterized`, and the Wave 2 GP protocols to already exist.

## GP Track Context
The GP design docs separate three concerns that should remain separate in these issues:

- guide family parameterization: `FullRankGuide`, `MeanFieldGuide`, `WhitenedGuide`, `NaturalGuide`, `DeltaGuide`
- inference strategy: direct ELBO optimization vs conjugate-computation/natural-gradient style updates
- likelihood family: Gaussian, Bernoulli, Poisson in this wave

The common sparse variational setup is:

$$q(u) = \mathcal{N}(m, S)$$

with ELBO

$$\mathcal{L} = \sum_n \mathbb{E}_{q(f_n)}[\log p(y_n \mid f_n)] - \mathrm{KL}[q(u)\|p(u)]$$

and later natural/site updates expressed through natural parameters rather than moment parameters.

## NN Track Context
Wave 3 NN work should stay focused on the foundational layer surface already named in the API docs:

- `DenseVariational`
- `DenseReparameterization`
- `DenseFlipout`
- `MCDropout`
- `NCPContinuousPerturb`
- `DenseNCP`
- `RandomFourierFeatures`
- `RandomKitchenSinks`

This is not yet the wave for advanced Edward2-style layers, sequence models, or spectral-normalization-based SNGP. Those appear in later grouped drafts.

## Random-Feature Baseline
The common approximation story that should be preserved in the issue text is:

$$k(x, x') \approx \phi(x)^\top \phi(x')$$

with feature map

$$\phi(x) = \sqrt{2/D}[\cos(Wx+b), \sin(Wx+b)]$$

This wave only needs the base random-feature story. Richer spectral families, variational frequencies, orthogonal features, and HSGP-style patterns belong in Wave 4.

## Explicit Non-Goals
Wave 3 should not absorb:

- EP, Laplace, posterior linearization, or quasi-Newton GP inference
- state-space kernels or temporal GP models
- pathwise posterior sampling
- multi-output GP models
- SNGP, DUE, or advanced Edward2 layers
- convolutional, recurrent, or attention families

---

## Problem / Request
Implement the first non-conjugate GP inference strategies and scalar likelihood families needed for Wave 3 sparse workflows.

## User Story
As a user fitting sparse or latent GPs with non-Gaussian observations, I want at least one direct variational path and one natural-gradient path plus reusable likelihood objects for common scalar data types.

## Motivation
Wave 3 sparse GP work needs a first real inference path plus likelihoods that match the common contracts. This issue turns the sparse guide catalog into runnable workflows.

## Proposed API
```python
class VariationalInference(eqx.Module):
    integrator: Integrator
    learning_rate: float = 0.01
    def elbo(self, ...): ...
    def update(self, posterior, likelihood, y): ...

class ConjugateVI(eqx.Module):
    integrator: Integrator
    damping: float = 1.0
    def compute_sites(self, likelihood, cavity_mean, cavity_var, y): ...
    def update(self, posterior, likelihood, y): ...

class GaussianLikelihood(eqx.Module): ...
class BernoulliLikelihood(eqx.Module): ...
class PoissonLikelihood(eqx.Module): ...
```

## Mathematical Notes
The issue should preserve the two central update stories.

ELBO path:

$$\mathcal{L} = \sum_n \mathbb{E}_{q(f_n)}[\log p(y_n\mid f_n)] - \mathrm{KL}[q(u)\|p(u)]$$

CVI site update:

$$\lambda_i^{(1)} = J_i - H_i m_i, \qquad \Lambda_i^{(2)} = -H_i$$

with expected derivatives $J_i = \mathbb{E}_q[
abla_f \log p(y_i\mid f_i)]$ and $H_i = \mathbb{E}_q[
abla_f^2 \log p(y_i\mid f_i)]$.

## References & Existing Code
- Design doc / spec: `design_docs/pyrox/features/gp/inference_strategies.md`, `design_docs/pyrox/features/gp/likelihoods.md`
- Existing code to compare against: the VI/CVI sections in `inference_strategies.md` and scalar likelihood formulas in `likelihoods.md`
- Related prior art: SVGP / CVI style non-conjugate GP inference

## Implementation Notes
- Implement `VariationalInference` and `ConjugateVI` against the guide and integrator surfaces.
- Implement Wave 3 likelihoods for Gaussian, Bernoulli, and Poisson observations with analytical gradients/Hessians where straightforward.
- Keep the issue bounded to scalar likelihoods and the two inference paths above; EP/Laplace and advanced likelihoods belong later.

## Implementation Steps
- [ ] Implement `VariationalInference` and `ConjugateVI` against the guide and integrator surfaces.
- [ ] Implement Wave 3 likelihoods for Gaussian, Bernoulli, and Poisson observations with analytical gradients/Hessians where straightforward.
- [ ] Keep the issue bounded to scalar likelihoods and the two inference paths above; EP/Laplace and advanced likelihoods belong later.

## Definition of Done
- Sparse/latent GP workflows can run with VI or CVI using the common likelihood surface.
- Wave 3 likelihood classes expose log-probability and derivative contracts compatible with later advanced inference methods.

## Testing
- Test ELL terms and derivatives for Gaussian, Bernoulli, and Poisson likelihoods.
- Test VI/CVI updates on tiny problems where trends are easy to verify numerically.

## Documentation
- Document which inference strategies and likelihoods are Wave 3-stable and which advanced methods are deferred.

## Relationships
- Parent epic: #26.
- Blocked by #28 and #29.
- Blocks Wave 4 temporal/pathwise sparse workflows.
